<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Boyi Guo (郭博逸), PhD - Projects</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./img/Boyi_Guo_20190924.jpeg" rel="icon" type="image/jpeg">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Boyi Guo (郭博逸), PhD</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://boyi-guo.com" rel="" target=""><i class="bi bi-house-door-fill" role="img" aria-label="Home">
</i> 
 <span class="menu-text"></span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link active" href="./projects.html" rel="" target="" aria-current="page">
 <span class="menu-text">Projects</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://www.overleaf.com/read/gdttrkmmwncb#4463ee" rel="" target="">
 <span class="menu-text">CV</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#scalable-machine-learning-models-for-precision-medicine" id="toc-scalable-machine-learning-models-for-precision-medicine" class="nav-link active" data-scroll-target="#scalable-machine-learning-models-for-precision-medicine">Scalable machine learning models for precision medicine</a>
  <ul class="collapse">
  <li><a href="#variable-and-functional-selection-using-spike-and-slab-lasso" id="toc-variable-and-functional-selection-using-spike-and-slab-lasso" class="nav-link" data-scroll-target="#variable-and-functional-selection-using-spike-and-slab-lasso">Variable and functional selection using spike-and-slab LASSO</a></li>
  <li><a href="#multivariate-random-forests-to-estimating-heterogeneous-treatment-effects" id="toc-multivariate-random-forests-to-estimating-heterogeneous-treatment-effects" class="nav-link" data-scroll-target="#multivariate-random-forests-to-estimating-heterogeneous-treatment-effects">Multivariate random forests to estimating heterogeneous treatment effects</a></li>
  </ul></li>
  <li><a href="#computational-omics" id="toc-computational-omics" class="nav-link" data-scroll-target="#computational-omics">Computational Omics</a>
  <ul class="collapse">
  <li><a href="#spatially-resolved-genomics" id="toc-spatially-resolved-genomics" class="nav-link" data-scroll-target="#spatially-resolved-genomics">Spatially-resolved genomics</a>
  <ul class="collapse">
  <li><a href="#computation-methods-processing-integrating-and-analyzing-spaital-genomics-data" id="toc-computation-methods-processing-integrating-and-analyzing-spaital-genomics-data" class="nav-link" data-scroll-target="#computation-methods-processing-integrating-and-analyzing-spaital-genomics-data">Computation methods processing, integrating and analyzing spaital genomics data</a></li>
  </ul></li>
  <li><a href="#microbiome" id="toc-microbiome" class="nav-link" data-scroll-target="#microbiome">Microbiome</a></li>
  </ul></li>
  <li><a href="#population-health" id="toc-population-health" class="nav-link" data-scroll-target="#population-health">Population Health</a>
  <ul class="collapse">
  <li><a href="#the-reasons-for-geographic-and-racial-differences-in-stroke" id="toc-the-reasons-for-geographic-and-racial-differences-in-stroke" class="nav-link" data-scroll-target="#the-reasons-for-geographic-and-racial-differences-in-stroke">The Reasons for Geographic and Racial Differences in Stroke</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Projects</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><img src="img/researchBanner.png" class="img-fluid" data-fig-align="center"> My research concentrates on developing statistically rigorous and computationally scalable machine-learning methods, as well as open-source software, that integrates population-scale multi-omics data to uncover functional mechanisms that explain disease heterogeneity.</p>
<hr>
<section id="scalable-machine-learning-models-for-precision-medicine" class="level1">
<h1>Scalable machine learning models for precision medicine</h1>
<p>Heterogeneity in treatment response and disease progression requires individualized health care informed by data-driven evidence. Deeply embedded in biomedical data, the key to successful precision medicine requires timely advances in statistically sounding machine learning models and AI tools that extract insights from the complex and high-dimensional biomedical data.</p>
<section id="variable-and-functional-selection-using-spike-and-slab-lasso" class="level2">
<h2 class="anchored" data-anchor-id="variable-and-functional-selection-using-spike-and-slab-lasso">Variable and functional selection using spike-and-slab LASSO</h2>
<div>
<div style="float: left; position: relative; top: 0px; padding: -10px;">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="img/SSL_GAM.png" class="img-fluid figure-img" style="width:80.0%"></p>
<figcaption class="figure-caption">Two-part spike-and-slab LASSO prior for smooth functions. (Guo et al., 2022)</figcaption>
</figure>
</div>
</div>
<p>I develop Bayesian statistical learning methods and scalable algorithms for high-dimensional data analysis, with the application to biomedical data that includes, but not limited to, population-level clinical abd -omics data. Specifically, I introduced new Bayesian hierarchical models, following the spike-and-slab LASSO framework, to select important nonlinear signals for high-dimensional data and predict disease risk of continuous, binary, count and survival outcomes. To overcome computational burden of training Bayesian models, I develop novel optimization-based model-fitting algorithms to improve computational scalability, and open-source software to make the analysis accessible to everyone.</p>
</div>
<ul>
<li><strong>Guo B</strong>, Jaeger BC, Rahman AKMF, Long DL, Yi N. (2022). <a href="https://onlinelibrary.wiley.com/doi/full/10.1002/sim.9483">Spike-and-slab LASSO generalized additive models and scalable algorithms for high-dimensional data analysis</a>. <em>Statistics in Medicine</em>.</li>
<li><strong>Guo B</strong>, Yi N. (2022). <a href="https://doi.org/10.48550/arXiv.2205.11600">A scalable and flexible Cox proportional hazard model for high-dimensional survival prediction and functional selection</a>. <em>arXiv</em>.</li>
<li><strong>Guo B</strong>, Yi N. (2022). <a href="https://doi.org/10.48550/arXiv.2207.02348">The R Package BHAM: Fast and scalable Bayesian hierarchical additive model for high-dimensional data</a>. <em>arXiv</em>.</li>
<li>Tang Z, Lei S, Zhang X, Yi Z, <strong>Guo B</strong>, Chen JY, Shen Y, Yi N. (2019). Gsslasso Cox: a Bayesian hierarchical model for predicting survival and detecting associated genes by incorporating pathway information. <em>BMC bioinformatics</em>.</li>
<li>Yi N, Tang Z, Zhang X, <strong>Guo B</strong>. (2019). BhGLM: Bayesian hierarchical GLMs and survival models, with applications to genomics and epidemiology. <em>Bioinformatics</em>.</li>
</ul>
</section>
<section id="multivariate-random-forests-to-estimating-heterogeneous-treatment-effects" class="level2">
<h2 class="anchored" data-anchor-id="multivariate-random-forests-to-estimating-heterogeneous-treatment-effects">Multivariate random forests to estimating heterogeneous treatment effects</h2>
<ul>
<li><strong>Guo B</strong>, Holscher H, Auvil L, Welge M, Bushell C, Novotny J, Baer D, Burd N, Khan N, Zhu R. (2021). <a href="https://doi.org/10.1007/s12561-021-09310-w">Estimating Heterogeneous Treatment Effect on Multivariate Responses Using Random Forests</a>. <em>Statistics in Biosciences.</em></li>
</ul>
<!-- , including unsupervised learning models to estimate heterogeneous treatment effect\cite{guo2021estimating}, count-based models to select differentiable features \cite{zhang2018negative, pendegraft2019bayesian, zhang2020zero} and high-dimensional predictive models to quantify risk \cite{guo2022spike, guo2022scalable, tang2019gsslasso}.  -->

<!--  Leveraging the complex omics data, e.g. microbiome or gene expression, we develop high-dimensional nonparametric models, including multivariate random forest and generalized additive models, to estimate individualized treatment effects and predict disease risks. These novel machine-learning models substantially reduce the analytic bias due to the parametric assumptions of existing methods and deliver easily interpretable evidence to inform decision-making and improve the understanding of disease etiology. -->
<!-- ### Bayesian spike-and-slab LASSO models for high-dimensional data analysis -->
<!-- ### Treatmeng Effect Prediction -->

<!-- 1. Novel nonparametric models to estimate individualized treatment effects and disease risk using high-dimensional omics data: Heterogeneous reaction to treatment or disease progression requires individualized medical decisions informed by data-driven evidence, improving the outlook of precision medicine. Leveraging the complex omics data, e.g. microbiome or gene expression, we develop high-dimensional nonparametric models, including multivariate random forest and generalized additive models, to estimate individualized treatment effects and predict disease risks. These novel machine-learning models substantially reduce the analytic bias due to the parametric assumptions of existing methods and deliver easily interpretable evidence to inform decision-making and improve the understanding of disease etiology. I led the design and development of these novel methods, benchmarked the proposed methods against the state-of-the-art methods and worked with interdisciplinary research teams to translate analytic results into novel findings. -->
<!-- a. Guo B, Yi N. A scalable and flexible Cox proportional hazards model for high-dimensional survival prediction and functional selection. arXiv. 2022 May. [Preprint.] DOI: 10.48550/ARXIV.2205.11600 -->
<!-- b. Guo B, Yi N. The R Package BHAM: Fast and Scalable Bayesian Hierarchical Additive Model for High-dimensional Data. arXiv. 2022 Jul. [Preprint.] DOI: 10.48550/ARXIV.2207.02348 -->
<!-- c. Guo B, Jaeger BC, Rahman AKMF, Long DL, Yi N. Spike-and-slab least absolute shrinkage and selection operator generalized additive models and scalable algorithms for high-dimensional data analysis. Stat Med. 2022 Sep 10;41(20):3899-3914. PubMed PMID: 35665524.  -->
<!-- d. Guo B, Holscher H, Auvil L, Welge M, Bushell C, Novotny J, Baer D, Burd N, Khan N, Zhu R. Estimating Heterogeneous Treatment Effect on Multivariate Responses Using Random Forests. Statistics in Biosciences. 2021 May 15; :-. Available from: https://link.springer.com/10.1007/s12561-021-09310-w DOI: 10.1007/s12561-021-09310-w -->
<!-- Scalable Bayesian hierarchical models to analyze high-dimensional microbiome data: Though enlightening human physiology, microbiome data poses significant analytic challenges, due to its sparse and high-dimensional nature. To analyze the complex microbiome data, we developed Bayesian hierarchical models to simultaneously investigate hundreds and thousands of microbes while rigorously controlling error rates and maintaining statistical power. To expedite the model fitting, we proposed computationally efficient optimization-based algorithms for the Bayesian models to improve scalability. During the methodology development, I led the benchmarking of the proposed methods against the state-of-the-art methods, designing and implementing numeric studies using synthetic data. -->
<!-- a. Pendegraft AH, Guo B, Yi N. Bayesian hierarchical negative binomial models for multivariable analyses with applications to human microbiome count data. PLoS One. 2019;14(8):e0220961. PubMed Central PMCID: PMC6706006.  -->
<!-- b. Zhang X, Guo B, Yi N. Zero-Inflated gaussian mixed models for analyzing longitudinal microbiome data. PLoS One. 2020;15(11):e0242073. PubMed Central PMCID: PMC7652264.  -->
<!-- c. Zhang X, Pei Y, Zhang L, Guo B, Pendegraft A, Zhuang W, Yi N. Negative Binomial Mixed Models for Analyzing Longitudinal Microbiome Data. Frontiers in Microbiology. 2018 DOI: 10.3389/fmicb.2018.01683 -->
<!-- ### Open-source software -->
<!-- Scientific software development translating cutting-edge statistical models to open-access interfaces: High-quality open-access software is indispensable for quantitative analysis, providing data-driven evidence to scientific hypotheses. Furthermore, with growing concerns about reproducibility, scientific software improves the standard by providing a well-documented well-designed interface, greatly reducing the technical burdens of conducting complex analysis. Leveraging my previous training in software engineering, I designed and implemented cutting-edge high-dimensional models in the R programming platform, translating theoretical statistical frameworks to open-access software. The software were validated and tested using both synthetic and real-world data. The developed software has been maintained. -->
<!--   a.   Guo B. A Reproducible Framework to Calculating Residential Segregation Indices With Decennial US Census Data. Zenodo, 2023 Jan. [Software.] DOI: 10.5281/ZENODO.7561874 -->
<!--   b.   Guo B, Zhu R. The R Package MOTE.RF: Estimating Heterogeneous Treatment Effect on Multivariate Responses Using Random Forests. Zenodo, 2023 Jan. [Software.] DOI: 10.5281/ZENODO.7561850 -->
<!--   c.   Guo B, Yi N. The R Package BHAM: Fast and Scalable Bayesian Hierarchical Additive Model for High-dimensional Data. arXiv. 2022 Jul. [Preprint.] DOI: 10.48550/ARXIV.2207.02348 -->
<!--   d.   Yi N, Tang Z, Zhang X, Guo B. BhGLM: Bayesian hierarchical GLMs and survival models, with applications to genomics and epidemiology. Bioinformatics. 2019 April 15; 35(8):1419-1421. DOI: 10.1093/bioinformatics/bty803 -->
<!-- ## Multiscale Data Analysis -->
<!-- ## Molecular and Cellular via Spatially resolved transcriptomics -->
</section>
</section>
<section id="computational-omics" class="level1">
<h1>Computational Omics</h1>
<section id="spatially-resolved-genomics" class="level2">
<h2 class="anchored" data-anchor-id="spatially-resolved-genomics">Spatially-resolved genomics</h2>
<p>The cortical layers of the human neocortex were classically defined by histological distinction of cell types according to size and shape. However, emerging single-cell and spatially-resolved transcriptomic technologies have facilitated the identification of molecularly defined cell populations and spatial domains that move beyond classic cell type definitions and cytoarchitectural boundaries. Assigning gene expression to distinct anatomical subdivisions and cell populations within the human brain improves the understanding of brain regions and how they contribute to brain disorders. I led the quantitative investigation of 30 Visium Samples and single-nucleus RNA seq collected from 19 brain samples, where integrative cell-cell communication identified FYN-EFNA5-EPHA5 signaling pathway associated with schizophrenia. I also developed a novel visualization framework that simultaneous display disparage variables following Gestalt principles, allowing to in situ visualize of differential expression across various spatial domains.</p>
<ul>
<li>Huuki-Myers L, Spangler A, Eagles N, Montgomery KD, Kwon SH, Guo B, Grant-Peters M, Divecha HR, Tippani M, Sriworarat C, Nguyen AB, Ravichandran P, Tran MN, Seyedian A, Hyde TM, Kleinman JE, Battle A, Page SC, Ryten M, Hicks SC, Martinowich K, Collado-Torres L, Maynard KR. Integrated single cell and unsupervised spatial transcriptomic analysis defines molecular anatomy of the human dorsolateral prefrontal cortex. bioRxiv. 2023 Feb 15; PubMed Central PMCID: PMC9949126.</li>
</ul>
<section id="computation-methods-processing-integrating-and-analyzing-spaital-genomics-data" class="level3">
<h3 class="anchored" data-anchor-id="computation-methods-processing-integrating-and-analyzing-spaital-genomics-data">Computation methods processing, integrating and analyzing spaital genomics data</h3>
<ul>
<li><strong>Guo B</strong>, Huuki-Myers LA, Grant-Peters M, Collado-Torres L, Hicks SC. <a href="https://academic.oup.com/bioinformaticsadvances/article/3/1/vbad179/7459905">escheR: Unified multi-dimensional visualizations with Gestalt principles</a>. <em>Bioinformatics Advances</em>.</li>
</ul>
</section>
</section>
<section id="microbiome" class="level2">
<h2 class="anchored" data-anchor-id="microbiome">Microbiome</h2>
</section>
</section>
<section id="population-health" class="level1">
<h1>Population Health</h1>
<section id="the-reasons-for-geographic-and-racial-differences-in-stroke" class="level2">
<h2 class="anchored" data-anchor-id="the-reasons-for-geographic-and-racial-differences-in-stroke">The Reasons for Geographic and Racial Differences in Stroke</h2>
<!-- Collaborative research studying health disparity in stroke and co-morbidities among racial groups and geographic locations: Black Americans under age 75 were more than twice as likely to die from stroke than were non-Hispanic white Americans, while people who lived in the South had a 40 percent higher risk of dying from stroke than their counterparts in other parts of the United States. Understanding the cause of the disparity is imperative to effective interventions and policy change. We analyzed data collected in the national cohort study, the Reasons for Geographic and Racial Differences in Stroke (REGARDS), to examine possible causal pathways and identify novel biomarkers to inform disease risk. I provided critical analytic support to medical doctors, epidemiologists and biologists. In the multi-center research group, I delivered talks and tutorials on reproducible analytic standards. -->
<!-- a. Kamin Mukaz D, Guo B, Long D, Judd S, Plante T, McClure L, Wolberg A, Zakai N, Howard G, Cushman M. D-dimer and the risk of hypertension: The REasons for Geographic And Racial Differences in Stroke Cohort Study. Research and Practice in Thrombosis and Haemostasis. 2023 January; 7(1):100016-. DOI: 10.1016/j.rpth.2022.100016 -->
<!-- b. Peper KM, Guo B, Leann Long D, Howard G, Carson AP, Howard VJ, Judd SE, Zakai NA, Cherrington A, Cushman M, Plante TB. C-reactive Protein and Racial Differences in Type 2 Diabetes Incidence: The REGARDS Study. J Clin Endocrinol Metab. 2022 May 17;107(6):e2523-e2531. PubMed Central PMCID: PMC9113826.  -->
<!-- c. Long DL, Guo B, McClure LA, Jaeger BC, Tison SE, Howard G, Judd SE, Howard VJ, Plante TB, Zakai NA, Koh I, Cheung KL, Cushman M. Biomarkers as MEDiators of racial disparities in risk factors (BioMedioR): Rationale, study design, and statistical considerations. Ann Epidemiol. 2022 Feb;66:13-19. PubMed Central PMCID: PMC8920757.  -->
<!-- d. Plante T, Long D, Guo B, Howard G, Carson A, Howard V, Judd S, Jenny N, Zakai N, Cushman M. C-Reactive Protein and Incident Hypertension in Black and White Americans in the REasons for Geographic And Racial Differences in Stroke (REGARDS) Cohort Study. American Journal of Hypertension. 2021 July 01; 34(7):698-706. DOI: 10.1093/ajh/hpaa215 -->


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>